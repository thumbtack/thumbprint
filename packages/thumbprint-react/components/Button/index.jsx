import React from 'react';
import PropTypes from 'prop-types';
import { Themed, Plain } from '../UIAction/index.jsx';

const getCommonProps = props => ({
    onClick: props.onClick,
    isDisabled: props.isDisabled,
    type: props.type,
    children: props.children,
    onMouseEnter: props.onMouseEnter,
    onMouseOver: props.onMouseOver,
    onFocus: props.onFocus,
    onMouseLeave: props.onMouseLeave,
    onBlur: props.onBlur,
    accessibilityLabel: props.accessibilityLabel,
    dataTest: props.dataTest,
});

const TextButton = React.forwardRef((props, ref) => (
    <Plain {...getCommonProps(props)} theme={props.theme} iconLeft={props.iconLeft} ref={ref} />
));

TextButton.propTypes = {
    /**
     * Contents displayed within the button.
     */
    children: PropTypes.node,
    /**
     * Icon from [Thumbprint Icons](/icons/) to render left of the text within `TextButton`.
     */
    iconLeft: PropTypes.node,
    /**
     * Visually and functionally disables the button. We discourage the use of this prop since it
     * is difficult to visually indicate that a link is disabled. Consider not rendering the
     * `TextButton` if it is not interactive.
     */
    isDisabled: PropTypes.bool,
    /**
     * Function that will run when the button is clicked on.
     */
    onClick: PropTypes.func,
    /**
     * Function that runs when the user hovers on the button.
     */
    onMouseEnter: PropTypes.func,
    /**
     * Function that runs when the user hovers on the button. Unlike `onMouseEnter`, `onMouseOver`
     * fires each time a child element receives focus.
     */
    onMouseOver: PropTypes.func,
    /**
     * Function that runs when the button receives focus.
     */
    onFocus: PropTypes.func,
    /**
     * Function that runs when the user hovers away from the button.
     */
    onMouseLeave: PropTypes.func,
    /**
     * Function that runs when the button loses focus.
     */
    onBlur: PropTypes.func,
    /**
     * Description of the button’s content. It is required if the button has an icon and no
     * descriptive text.
     */
    accessibilityLabel: PropTypes.string,
    /**
     * Controls the button’s background, text, and border color.
     * `inherit` will make the button inherit `color` from its parent.
     */
    theme: PropTypes.oneOf(['primary', 'secondary', 'tertiary', 'inherit']),
    /**
     * Button’s of type `submit` will submit a form when used within a `form` element.
     */
    type: PropTypes.oneOf(['button', 'submit']),
    /**
     * A selector hook into the React component for use in automated testing environments.
     */
    dataTest: PropTypes.string,
};

TextButton.defaultProps = {
    children: undefined,
    isDisabled: false,
    iconLeft: null,
    theme: 'primary',
    onClick: null,
    onMouseEnter: undefined,
    onMouseOver: undefined,
    onFocus: undefined,
    onMouseLeave: undefined,
    onBlur: undefined,
    type: 'button',
    accessibilityLabel: undefined,
    dataTest: undefined,
};

const Button = React.forwardRef((props, ref) => (
    <Themed
        {...getCommonProps(props)}
        icon={props.icon}
        isLoading={props.isLoading}
        theme={props.theme}
        size={props.size}
        width={props.width}
        ref={ref}
    />
));

Button.propTypes = {
    /**
     * Contents displayed within the button.
     */
    children: PropTypes.node,
    /**
     * Visually and functionally disables the button.
     */
    isDisabled: PropTypes.bool,
    /**
     * Boolean determining whether the button is in a loading state. When `true` the text will
     * we replaced with a loading animation and interaction will be disabled.
     */
    isLoading: PropTypes.bool,
    /**
     * Icon from [Thumbprint Icons](/icons/) to render within the button. It must be one of the
     * `small` icons.
     */
    icon: PropTypes.node,
    /**
     * Description of the button’s content. It is required if the button has an icon and no
     * descriptive text.
     */
    accessibilityLabel: PropTypes.string,
    /**
     * Button’s of type `submit` will submit a form when used within a `form` element.
     */
    type: PropTypes.oneOf(['button', 'submit']),
    /**
     * Function that will run when the button is clicked on.
     */
    onClick: PropTypes.func,
    /**
     * Function that runs when the user hovers on the button.
     */
    onMouseEnter: PropTypes.func,
    /**
     * Function that runs when the user hovers on the button. Unlike `onMouseEnter`, `onMouseOver`
     * fires each time a child element receives focus.
     */
    onMouseOver: PropTypes.func,
    /**
     * Function that runs when the button receives focus.
     */
    onFocus: PropTypes.func,
    /**
     * Function that runs when the user hovers away from the button.
     */
    onMouseLeave: PropTypes.func,
    /**
     * Function that runs when the button loses focus.
     */
    onBlur: PropTypes.func,
    /**
     * Controls the button’s background, text, and border color.
     */
    theme: PropTypes.oneOf(['primary', 'secondary', 'caution', 'solid']),
    /**
     * Changes the button's `line-height`, `padding`, `border-radius`, and `font-size`.
     */
    size: PropTypes.oneOf(['small', 'large']),
    /**
     * `Button` components are as wide as the content that is passed in. The `full` option will
     * expand the width to `100%` on all screens. `full-below-small` will expand the width to 100%
     * on devices smaller than [our `small` breakpoint](/tokens/#section-breakpoint).
     */
    width: PropTypes.oneOf(['auto', 'full', 'full-below-small']),
    /**
     * A selector hook into the React component for use in automated testing environments.
     */
    dataTest: PropTypes.string,
};

Button.defaultProps = {
    children: undefined,
    type: 'button',
    isDisabled: false,
    isLoading: false,
    icon: undefined,
    onClick: undefined,
    onMouseEnter: undefined,
    onMouseOver: undefined,
    onFocus: undefined,
    onMouseLeave: undefined,
    onBlur: undefined,
    theme: 'primary',
    size: 'large',
    width: 'auto',
    accessibilityLabel: undefined,
    dataTest: undefined,
};

export default Button;
export { TextButton };
