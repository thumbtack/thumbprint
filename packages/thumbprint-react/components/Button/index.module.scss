@import '~@thumbtack/thumbprint-tokens/dist/scss/_index';
@import '~@thumbtack/thumbprint-scss/mixins';

$border-width: 2px;

.root {
    border: 2px solid;
    border-radius: 4px;
    font-weight: $tp-font-weight__bold;
    // The designs have an inset 2px border. The subtraction accounts for that.
    padding: 0 ($tp-space__4 - $border-width);
    // Used to absolutely position the `LoaderDots`
    position: relative;
    display: inline-block;
    user-select: none;
}

.buttonWidthFull {
    width: 100%;
}

.buttonWidthFullBelowSmall {
    @include tp-respond-below($tp-breakpoint__small) {
        width: 100%;
    }
}

.buttonLarge,
.buttonMedium {
    min-height: (52px - ($border-width * 2));
}

.buttonLarge {
    // This doesn't use one of our system font sizes because the text size needs to be just large
    // enough so that it is accessible on our blue background.
    font-size: 19px;
}

.buttonMedium {
    font-size: $tp-font__body__1__size;

    @include tp-respond-below($tp-breakpoint__small) {
        min-height: 48px;
    }
}

.buttonSmall {
    font-size: $tp-font__body__2__size;
    min-height: 36px - ($border-width * 2);
}

.primaryButton {
    color: $tp-color__white;
    border-color: $tp-color__blue;
    background-color: $tp-color__blue;

    &[disabled] {
        border-color: $tp-color__blue-300;
        background-color: $tp-color__blue-300;
    }
}

.secondaryButtonEmphasisHigh {
    color: $tp-color__white;
    border-color: $tp-color__black;
    background-color: $tp-color__black;
}

.secondaryButtonEmphasisLow {
    color: $tp-color__black;
    border-color: $tp-color__white;
    background-color: $tp-color__white;
}

.secondaryButtonBlackBorder {
    border-color: $tp-color__black;

    &[disabled] {
        background-color: $tp-color__gray;
        border-color: $tp-color__gray;
    }
}

.secondaryButtonWhiteBorder {
    border-color: $tp-color__white;

    &[disabled] {
        color: $tp-color__gray;
    }
}

.secondaryButtonGrayBorder {
    border-color: $tp-color__gray;

    &[disabled] {
        color: $tp-color__gray;
    }
}

.tertiaryButton {
    border-color: $tp-color__white;
    background-color: $tp-color__white;
    color: $tp-color__black-300;
}

.cautionButton[disabled] {
    border-color: $tp-color__gray;
    background-color: $tp-color__red-100;
    color: $tp-color__red-200;
}

.cautionButtonEmphasisHigh {
    background-color: $tp-color__red-500;
    color: $tp-color__white;
    border-color: $tp-color__red-500;
}

.cautionButtonEmphasisLow {
    background-color: $tp-color__white;
    color: $tp-color__red-500;
    border-color: $tp-color__gray;
}

.flexWrapper {
    // This wraps the content since `button`s are not able to use `display: flex`.
    display: flex;
    align-items: center;
    justify-content: center;
}
