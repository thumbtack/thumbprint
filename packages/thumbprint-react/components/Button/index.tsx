import React from 'react';
import { Themed, Plain } from '../UIAction/index';

interface CommonProps {
    children?: React.ReactNode;
    iconLeft?: React.ReactNode;
    iconRight?: React.ReactNode;
    isDisabled?: boolean;
    onClick?: () => void;
    onMouseEnter?: () => void;
    onMouseOver?: () => void;
    onFocus?: () => void;
    onMouseLeave?: () => void;
    onBlur?: () => void;
    theme?: 'primary' | 'secondary' | 'tertiary' | 'inherit';
    type?: 'button' | 'submit';
    dataTest?: string;
    accessibilityLabel?: string;
}

const getCommonProps = (props: CommonProps): CommonProps => ({
    onClick: props.onClick,
    isDisabled: props.isDisabled,
    type: props.type,
    children: props.children,
    onMouseEnter: props.onMouseEnter,
    onMouseOver: props.onMouseOver,
    onFocus: props.onFocus,
    onMouseLeave: props.onMouseLeave,
    onBlur: props.onBlur,
    accessibilityLabel: props.accessibilityLabel,
    dataTest: props.dataTest,
});

const TextButton = React.forwardRef<HTMLButtonElement, TextButtonPropTypes>(
    (
        {
            children,
            iconLeft,
            iconRight,
            isDisabled = false,
            onClick,
            onMouseEnter,
            onMouseOver,
            onFocus,
            onMouseLeave,
            onBlur,
            accessibilityLabel,
            theme = 'primary',
            type = 'button',
            dataTest,
        }: TextButtonPropTypes,
        ref,
    ): JSX.Element => (
        <Plain
            {...getCommonProps({
                onClick,
                isDisabled,
                type,
                children,
                onMouseEnter,
                onMouseOver,
                onFocus,
                onMouseLeave,
                onBlur,
                accessibilityLabel,
                dataTest,
            })}
            theme={theme}
            iconLeft={iconLeft}
            iconRight={iconRight}
            ref={ref}
        />
    ),
);

interface TextButtonPropTypes {
    /**
     * Contents displayed within the button.
     */
    children?: React.ReactNode;
    /**
     * Icon from [Thumbprint Icons](/icons/) to render left of the text within `TextButton`.
     */
    iconLeft?: React.ReactNode;
    /**
     * Icon from [Thumbprint Icons](/icons/) to render right of the text within `TextButton`.
     */
    iconRight?: React.ReactNode;
    /**
     * Visually and functionally disables the button. We discourage the use of this prop since it
     * is difficult to visually indicate that a link is disabled. Consider not rendering the
     * `TextButton` if it is not interactive.
     */
    isDisabled?: boolean;
    /**
     * Function that will run when the button is clicked on.
     */
    onClick?: () => void;
    /**
     * Function that runs when the user hovers on the button.
     */
    onMouseEnter?: () => void;
    /**
     * Function that runs when the user hovers on the button. Unlike `onMouseEnter`, `onMouseOver`
     * fires each time a child element receives focus.
     */
    onMouseOver?: () => void;
    /**
     * Function that runs when the button receives focus.
     */
    onFocus?: () => void;
    /**
     * Function that runs when the user hovers away from the button.
     */
    onMouseLeave?: () => void;
    /**
     * Function that runs when the button loses focus.
     */
    onBlur?: () => void;
    /**
     * Description of the button’s content. It is required if the button has an icon and no
     * descriptive text.
     */
    accessibilityLabel?: string;
    /**
     * Controls the button’s background, text, and border color.
     * `inherit` will make the button inherit `color` from its parent.
     */
    theme?: 'primary' | 'secondary' | 'tertiary' | 'inherit';
    /**
     * Button’s of type `submit` will submit a form when used within a `form` element.
     */
    type?: 'button' | 'submit';
    /**
     * A selector hook into the React component for use in automated testing environments.
     */
    dataTest?: string;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonPropTypes>(
    (
        {
            children,
            isDisabled = false,
            isLoading = false,
            icon,
            accessibilityLabel,
            type = 'button',
            onClick,
            onMouseEnter,
            onMouseOver,
            onFocus,
            onMouseLeave,
            onBlur,
            theme = 'primary',
            size = 'large',
            width = 'auto',
            dataTest,
        }: ButtonPropTypes,
        ref,
    ): JSX.Element => (
        <Themed
            {...getCommonProps({
                onClick,
                isDisabled,
                type,
                children,
                onMouseEnter,
                onMouseOver,
                onFocus,
                onMouseLeave,
                onBlur,
                accessibilityLabel,
                dataTest,
            })}
            icon={icon}
            isLoading={isLoading}
            theme={theme}
            size={size}
            width={width}
            ref={ref}
        />
    ),
);

interface ButtonPropTypes {
    /**
     * Contents displayed within the button.
     */
    children?: React.ReactNode;
    /**
     * Visually and functionally disables the button.
     */
    isDisabled?: boolean;
    /**
     * Boolean determining whether the button is in a loading state. When `true` the text will
     * we replaced with a loading animation and interaction will be disabled.
     */
    isLoading?: boolean;
    /**
     * Icon from [Thumbprint Icons](/icons/) to render within the button. It must be one of the
     * `small` icons.
     */
    icon?: React.ReactNode;
    /**
     * Description of the button’s content. It is required if the button has an icon and no
     * descriptive text.
     */
    accessibilityLabel?: string;
    /**
     * Button’s of type `submit` will submit a form when used within a `form` element.
     */
    type?: 'button' | 'submit';
    /**
     * Function that will run when the button is clicked on.
     */
    onClick?: () => void;
    /**
     * Function that runs when the user hovers on the button.
     */
    onMouseEnter?: () => void;
    /**
     * Function that runs when the user hovers on the button. Unlike `onMouseEnter`, `onMouseOver`
     * fires each time a child element receives focus.
     */
    onMouseOver?: () => void;
    /**
     * Function that runs when the button receives focus.
     */
    onFocus?: () => void;
    /**
     * Function that runs when the user hovers away from the button.
     */
    onMouseLeave?: () => void;
    /**
     * Function that runs when the button loses focus.
     */
    onBlur?: () => void;
    /**
     * Controls the button’s background, text, and border color.
     */
    theme?: 'primary' | 'secondary' | 'tertiary' | 'caution' | 'solid';
    /**
     * Changes the button's `line-height`, `padding`, `border-radius`, and `font-size`.
     */
    size?: 'small' | 'large';
    /**
     * `Button` components are as wide as the content that is passed in. The `full` option will
     * expand the width to `100%` on all screens. `full-below-small` will expand the width to 100%
     * on devices smaller than [our `small` breakpoint](/tokens/#section-breakpoint).
     */
    width?: 'auto' | 'full' | 'full-below-small';
    /**
     * A selector hook into the React component for use in automated testing environments.
     */
    dataTest?: string;
}

export default Button;
export { TextButton };
