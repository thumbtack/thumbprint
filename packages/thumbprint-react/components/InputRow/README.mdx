---
package: '@thumbtack/thumbprint-react'
kit: input-row/index.mdx
platform: react
url: /components/input-row/react/
mdxType: componentApi
---

## `InputRow` with an `Input` and `Button`

The `widthRatios` prop in this example tells the input to take up as much space as possible.

```jsx
<InputRow widthRatios={[1, null]}>
    <Input placeholder="Enter a zip code" onChange={() => {}} />
    <Button>Find a pro</Button>
</InputRow>
```

## Custom components within `InputRow`

The `Input` and `Button` components use [React's Context API](https://reactjs.org/docs/context.html) to remove their rounded borders when used within an `InputRow`.

The `InputRow` exports its React context as `InputRowContext`. This makes it possible to use custom components within `InputRow`.

To use `InputRowContext`, start by importing it within your component:

```js
import { InputRowContext } from '@thumbtack/thumbprint-react';
```

Within your component's JSX, you'll use `<InputRowContext.Consumer>` to gain access to three booleans: `isWithinInputRow`, `isFirstInputRowChild`, `isLastInputRowChild`.

Here's an example:

```jsx shouldRender=false
<InputRowContext.Consumer>
    {({ isWithinInputRow, isFirstInputRowChild, isLastInputRowChild }) => <div />}
</InputRowContext.Consumer>
```

We recommend doing the following with these booleans:

-   Remove your component's right border when `isWithinInputRow && !isLastInputRowChild`.
-   Remove left rounded corners when `isWithinInputRow && !isFirstInputRowChild`.
-   Remove right rounded corners when `isWithinInputRow && !isLastInputRowChild`.

You can [learn more about React's Context API](https://reactjs.org/docs/context.html) in the React documentation.
