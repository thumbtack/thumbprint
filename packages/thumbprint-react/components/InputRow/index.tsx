import React from 'react';
import warning from 'warning';
import styles from './index.module.scss';

interface PropTypes {
    /**
     * Form fields to display.
     */
    children: React.ReactNode;
    /**
     * An array of numbers (or `null`) that reflect the ratio of the widths of the `children`.
     *
     * Examples:
     *
     * * `[1, null]`: the first child takes up as much space as possible
     * * `[1, 1, 1]`: splits the width evenly between three elements
     * * `[1, 2, 1]`: the first and third child have the same width. The middle child is twice as
     * wide as its siblings.
     *
     * The length of the array should be the same as the number of children.
     */
    widthRatios?: (number | null)[];
    /**
     * A selector hook into the React component for use in automated testing environments.
     */
    dataTestId?: string;
    /**
     * A selector hook into the React component for use in automated testing environments.
     * @deprecated Deprecated in favor of the `dataTestId` prop
     */
    dataTest?: string;
}

// Components that import `InputRowContext` from this package will have access to this object. They
// will then be able to know when their component is being used within an `InputRow` as well as
// whether it is the first of last child.
const InputRowContext = React.createContext({
    isWithinInputRow: false,
    isFirstInputRowChild: false,
    isLastInputRowChild: false,
});

export { InputRowContext };

export default function InputRow({
    widthRatios,
    children,
    dataTest,
    dataTestId,
}: PropTypes): JSX.Element {
    const widthLength = widthRatios && widthRatios.length;
    const numChildren = React.Children.count(children);

    warning(
        !widthLength || widthLength === numChildren,
        `The length of prop \`width\` (${widthLength}) must match the number of children (${numChildren}) passed to \`InputRow\`.`,
    );

    warning(
        numChildren > 1,
        `The \`InputRow\` component expects more than one child within the \`children\` prop. ${numChildren} received.`,
    );

    return (
        <div className={styles.root} data-test={dataTest} data-testid={dataTestId}>
            {React.Children.map(children, (child, i) => (
                <div style={{ flex: widthRatios ? `${widthRatios[i]}` : undefined }}>
                    <InputRowContext.Provider
                        value={{
                            isWithinInputRow: true,
                            isFirstInputRowChild: i === 0,
                            isLastInputRowChild: i + 1 === numChildren,
                        }}
                    >
                        {child}
                    </InputRowContext.Provider>
                </div>
            ))}
        </div>
    );
}
