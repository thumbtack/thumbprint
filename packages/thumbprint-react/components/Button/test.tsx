import React from 'react';
import { mount } from 'enzyme';
import Button, { TextButton } from './index';

test('renders children passed in', (): void => {
    const testComponent = (Component: React.ElementType): void => {
        const wrapper = mount(<Component>Goose</Component>);
        expect(wrapper.text()).toEqual('Goose');
        expect(wrapper).toMatchSnapshot();
    };

    testComponent(Button);
    testComponent(TextButton);
});

test('renders button of type `button` by default', (): void => {
    const testComponent = (Component: React.ElementType): void => {
        const wrapper = mount(<Component>Goose</Component>);
        expect(wrapper.find('button').prop('type')).toEqual('button');
        expect(wrapper).toMatchSnapshot();
    };

    testComponent(Button);
    testComponent(TextButton);
});

test('renders button of type `submit` when `type` is set', (): void => {
    const testComponent = (Component: React.ElementType): void => {
        const wrapper = mount(<Component type="submit">Goose</Component>);
        expect(wrapper.find('button').prop('type')).toEqual('submit');
        expect(wrapper).toMatchSnapshot();
    };

    testComponent(Button);
    testComponent(TextButton);
});

test('`onClick` runs when button is clicked on', (): void => {
    const testComponent = (Component: React.ElementType): void => {
        const onClick = jest.fn();
        const wrapper = mount(<Component onClick={onClick}>Goose</Component>);

        expect(onClick).toHaveBeenCalledTimes(0);
        wrapper.simulate('click');
        expect(onClick).toHaveBeenCalledTimes(1);
    };

    testComponent(Button);
    testComponent(TextButton);
});

test('`onMouseOver` runs when button is moused over', (): void => {
    const testComponent = (Component: React.ElementType): void => {
        const onMouseOver = jest.fn();
        // eslint-disable-next-line jsx-a11y/mouse-events-have-key-events
        const wrapper = mount(<Component onMouseOver={onMouseOver}>Goose</Component>);

        expect(onMouseOver).toHaveBeenCalledTimes(0);
        wrapper.simulate('mouseover');
        expect(onMouseOver).toHaveBeenCalledTimes(1);
    };

    testComponent(Button);
    testComponent(TextButton);
});

test('renders icon that is passed in', (): void => {
    const wrapper = mount(<Button icon={<svg>Duck</svg>}>Goose</Button>);
    expect(wrapper.find('svg')).toHaveLength(1);
    expect(wrapper.text()).toBe('DuckGoose');
    expect(wrapper).toMatchSnapshot();
});

test('adds `dataTest` prop', (): void => {
    const wrapperA = mount(<Button dataTest="Duck">Goose</Button>);
    expect(wrapperA.find('button').prop('data-test')).toEqual('Duck');
    expect(wrapperA).toMatchSnapshot();

    const wrapperB = mount(<TextButton dataTest="Duck">Goose</TextButton>);
    expect(wrapperB.find('button').prop('data-test')).toEqual('Duck');
    expect(wrapperB).toMatchSnapshot();
});
