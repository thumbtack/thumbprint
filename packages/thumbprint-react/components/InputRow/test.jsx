import React from 'react';
import { mount } from 'enzyme';
import Input from '../Input/index';
import Button from '../Button/index';
import InputRow from './index';

test('renders all children', () => {
    const wrapper = mount(
        <InputRow>
            <button type="button" />
            <button type="button" />
            <button type="button" />
        </InputRow>,
    );
    expect(wrapper.find('button')).toHaveLength(3);
    expect(wrapper).toMatchSnapshot();
});

test('sets width ratios when they are all numbers', () => {
    const wrapper = mount(
        <InputRow widthRatios={[1, 1, 1]}>
            <button type="button" />
            <button type="button" />
            <button type="button" />
        </InputRow>,
    );
    expect(wrapper).toMatchSnapshot();
});

test('sets width ratios even if some are `null`', () => {
    const wrapper = mount(
        <InputRow widthRatios={[1, null, null]}>
            <button type="button" />
            <button type="button" />
            <button type="button" />
        </InputRow>,
    );
    expect(wrapper).toMatchSnapshot();
});

test('works with `Input` and `Button`', () => {
    // Test out different ordering because the components should behave differently dpeending on
    // their orders. Some components, such as `Button`, set `width: 100%` if they are used within
    // an `InputRow`.
    const wrapperA = mount(
        <InputRow>
            <Input onChange={jest.fn} />
            <Button>Duck</Button>
        </InputRow>,
    );

    const wrapperB = mount(
        <InputRow>
            <Button>Duck</Button>
            <Input onChange={jest.fn} />
        </InputRow>,
    );

    const wrapperC = mount(
        <InputRow>
            <Button>Duck</Button>
            <Input onChange={jest.fn} />
        </InputRow>,
    );

    const wrapperD = mount(
        <InputRow>
            <Input onChange={jest.fn} />
            <Button>Duck</Button>
        </InputRow>,
    );

    expect(wrapperA).toMatchSnapshot();
    expect(wrapperB).toMatchSnapshot();
    expect(wrapperC).toMatchSnapshot();
    expect(wrapperD).toMatchSnapshot();
});

test('adds `dataTest` prop', () => {
    const wrapper = mount(
        <InputRow dataTest="Duck">
            <button type="button" />
            <button type="button" />
            <button type="button" />
        </InputRow>,
    );
    expect(wrapper.find('div[data-test="Duck"]')).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
});
