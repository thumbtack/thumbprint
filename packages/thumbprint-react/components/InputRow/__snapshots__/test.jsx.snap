// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds \`dataTest\` prop 1`] = `
<InputRow
  dataTest="Duck"
>
  <div
    className="root"
    data-test="Duck"
  >
    <div
      key=".0"
      style={
        Object {
          "flex": undefined,
        }
      }
    >
      <button
        type="button"
      />
    </div>
    <div
      key=".1"
      style={
        Object {
          "flex": undefined,
        }
      }
    >
      <button
        type="button"
      />
    </div>
    <div
      key=".2"
      style={
        Object {
          "flex": undefined,
        }
      }
    >
      <button
        type="button"
      />
    </div>
  </div>
</InputRow>
`;

exports[`renders all children 1`] = `
<InputRow>
  <div
    className="root"
  >
    <div
      key=".0"
      style={
        Object {
          "flex": undefined,
        }
      }
    >
      <button
        type="button"
      />
    </div>
    <div
      key=".1"
      style={
        Object {
          "flex": undefined,
        }
      }
    >
      <button
        type="button"
      />
    </div>
    <div
      key=".2"
      style={
        Object {
          "flex": undefined,
        }
      }
    >
      <button
        type="button"
      />
    </div>
  </div>
</InputRow>
`;

exports[`sets width ratios even if some are \`null\` 1`] = `
<InputRow
  widthRatios={
    Array [
      1,
      null,
      null,
    ]
  }
>
  <div
    className="root"
  >
    <div
      key=".0"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <button
        type="button"
      />
    </div>
    <div
      key=".1"
      style={
        Object {
          "flex": null,
        }
      }
    >
      <button
        type="button"
      />
    </div>
    <div
      key=".2"
      style={
        Object {
          "flex": null,
        }
      }
    >
      <button
        type="button"
      />
    </div>
  </div>
</InputRow>
`;

exports[`sets width ratios when they are all numbers 1`] = `
<InputRow
  widthRatios={
    Array [
      1,
      1,
      1,
    ]
  }
>
  <div
    className="root"
  >
    <div
      key=".0"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <button
        type="button"
      />
    </div>
    <div
      key=".1"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <button
        type="button"
      />
    </div>
    <div
      key=".2"
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <button
        type="button"
      />
    </div>
  </div>
</InputRow>
`;

exports[`works with \`Input\` and \`Button\` 1`] = `
<InputRow>
  <div
    className="root"
  >
    <div
      key=".0"
      style={
        Object {
          "flex": undefined,
        }
      }
    >
      <Input
        onChange={[Function]}
      >
        <div
          className="root rootUiStateDefault"
        >
          <input
            autoFocus={false}
            className="input inputSizeLarge"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
            readOnly={false}
            required={false}
            type="text"
            value=""
          />
          <div
            className="inputStyles inputStylesRoundedBordersLeft inputStylesHasNoRightBorder inputStylesUiStateDefault"
          />
        </div>
      </Input>
    </div>
    <div
      key=".1"
      style={
        Object {
          "flex": undefined,
        }
      }
    >
      <ForwardRef>
        <ForwardRef
          isDisabled={false}
          isLoading={false}
          size="large"
          theme="primary"
          type="button"
          width="auto"
        >
          <button
            className="themedButton themedButtonRoundedBordersRight themedButtonThemePrimary themedButtonWidthFull"
            disabled={false}
            type="button"
          >
            <span
              className="flexWrapper flexWrapperSizeLarge"
            >
              Duck
            </span>
          </button>
        </ForwardRef>
      </ForwardRef>
    </div>
  </div>
</InputRow>
`;

exports[`works with \`Input\` and \`Button\` 2`] = `
<InputRow>
  <div
    className="root"
  >
    <div
      key=".0"
      style={
        Object {
          "flex": undefined,
        }
      }
    >
      <ForwardRef>
        <ForwardRef
          isDisabled={false}
          isLoading={false}
          size="large"
          theme="primary"
          type="button"
          width="auto"
        >
          <button
            className="themedButton themedButtonRoundedBordersLeft themedButtonHasNoRightBorder themedButtonThemePrimary themedButtonWidthFull"
            disabled={false}
            type="button"
          >
            <span
              className="flexWrapper flexWrapperSizeLarge"
            >
              Duck
            </span>
          </button>
        </ForwardRef>
      </ForwardRef>
    </div>
    <div
      key=".1"
      style={
        Object {
          "flex": undefined,
        }
      }
    >
      <Input
        onChange={[Function]}
      >
        <div
          className="root rootUiStateDefault"
        >
          <input
            autoFocus={false}
            className="input inputSizeLarge"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
            readOnly={false}
            required={false}
            type="text"
            value=""
          />
          <div
            className="inputStyles inputStylesRoundedBordersRight inputStylesUiStateDefault"
          />
        </div>
      </Input>
    </div>
  </div>
</InputRow>
`;

exports[`works with \`Input\` and \`Button\` 3`] = `
<InputRow>
  <div
    className="root"
  >
    <div
      key=".0"
      style={
        Object {
          "flex": undefined,
        }
      }
    >
      <ForwardRef>
        <ForwardRef
          isDisabled={false}
          isLoading={false}
          size="large"
          theme="primary"
          type="button"
          width="auto"
        >
          <button
            className="themedButton themedButtonRoundedBordersLeft themedButtonHasNoRightBorder themedButtonThemePrimary themedButtonWidthFull"
            disabled={false}
            type="button"
          >
            <span
              className="flexWrapper flexWrapperSizeLarge"
            >
              Duck
            </span>
          </button>
        </ForwardRef>
      </ForwardRef>
    </div>
    <div
      key=".1"
      style={
        Object {
          "flex": undefined,
        }
      }
    >
      <Input
        onChange={[Function]}
      >
        <div
          className="root rootUiStateDefault"
        >
          <input
            autoFocus={false}
            className="input inputSizeLarge"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
            readOnly={false}
            required={false}
            type="text"
            value=""
          />
          <div
            className="inputStyles inputStylesRoundedBordersRight inputStylesUiStateDefault"
          />
        </div>
      </Input>
    </div>
  </div>
</InputRow>
`;

exports[`works with \`Input\` and \`Button\` 4`] = `
<InputRow>
  <div
    className="root"
  >
    <div
      key=".0"
      style={
        Object {
          "flex": undefined,
        }
      }
    >
      <Input
        onChange={[Function]}
      >
        <div
          className="root rootUiStateDefault"
        >
          <input
            autoFocus={false}
            className="input inputSizeLarge"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
            readOnly={false}
            required={false}
            type="text"
            value=""
          />
          <div
            className="inputStyles inputStylesRoundedBordersLeft inputStylesHasNoRightBorder inputStylesUiStateDefault"
          />
        </div>
      </Input>
    </div>
    <div
      key=".1"
      style={
        Object {
          "flex": undefined,
        }
      }
    >
      <ForwardRef>
        <ForwardRef
          isDisabled={false}
          isLoading={false}
          size="large"
          theme="primary"
          type="button"
          width="auto"
        >
          <button
            className="themedButton themedButtonRoundedBordersRight themedButtonThemePrimary themedButtonWidthFull"
            disabled={false}
            type="button"
          >
            <span
              className="flexWrapper flexWrapperSizeLarge"
            >
              Duck
            </span>
          </button>
        </ForwardRef>
      </ForwardRef>
    </div>
  </div>
</InputRow>
`;
